pkg/kube-metrics/collector.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubemetrics
+
+import (
+	"context"
+
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	kcollector "k8s.io/kube-state-metrics/pkg/collector"
+	"k8s.io/kube-state-metrics/pkg/metric"
+	metricsstore "k8s.io/kube-state-metrics/pkg/metrics_store"
+)
+
+// NewCollector returns a collection of metrics in the namespaces provided, per the api/kind resource.
+// The metrics are registered in the custom generateStore function that needs to be defined.
+// Current operators namespace will be added to the passed namespaces.
+func NewCollector(uc *Client,
+	namespaces []string,
+	api string,
+	kind string,
+	generateStore []metric.FamilyGenerator) ([]*kcollector.Collector, error) {
+	var collectors []*kcollector.Collector
+	// Detect in which namespace the operator is deployed in.
+	ns, err := k8sutil.GetOperatorNamespace()
+	if err != nil {
+		return nil, err
+	}
+	namespaces = append(namespaces, ns)
+	namespaces = deduplicateNamespaces(namespaces)
+
+	// Get collector per namespace.
+	for _, ns := range namespaces {
+		dclient, err := uc.ClientFor(api, kind, ns)
+		if err != nil {
+			return nil, err
+		}
+		composedMetricGenFuncs := metric.ComposeMetricGenFuncs(generateStore)
+		headers := metric.ExtractMetricFamilyHeaders(generateStore)
+		store := metricsstore.NewMetricsStore(headers, composedMetricGenFuncs)
+		reflectorPerNamespace(context.TODO(), dclient, &unstructured.Unstructured{}, store, ns)
+		collector := kcollector.NewCollector(store)
+		collectors = append(collectors, collector)
+	}
+
+	return collectors, nil
+}
+
+func deduplicateNamespaces(ns []string) []string {
```

Comment:
Style suggestion:
```go
func deduplicateNamespaces(ns []string) (list []string) {
	keys := make(map[string]struct{})
	for _, entry := range ns {
		if _, ok := keys[entry]; !ok {
			keys[entry] = struct{}{}
			list = append(list, entry)
		}
	}
	return list
}
```
Empty `struct{}`'s are 0-width so cost less to use in a set. Metrics setup isn't a performance bottleneck so these changes aren't necessary, just a suggestion.

Reply:
While I agree, it's a valid suggestion, I think the code doesn't read as clearly as the original implementation. But if everyone else wants to do the above, will be happy to change it.

Reply:
Both read the same to me. I've seen enough `map[string]struct{}` declarations in Go code now that I recognize that as a set of strings, and that's what I use by default because of the memory efficiency.